asf_core_data/pipeline/data_joining/merge_proc_datasets.py

```diff
@@ -0,0 +1,206 @@
+        print("EPC", epc_df.shape)
+        print("MCS", mcs_df.shape)
+        print("MCS (EPC matched)", mcs_matched_df.shape)
+        print("MCS (EPC unmatched)", mcs_non_matched_df.shape)
+
+    epc_mcs_df = pd.merge(
+        epc_df,
+        mcs_matched_df,
+        on=["UPRN", "EPC_AVAILABLE", "MCS_AVAILABLE"],
+        how="outer",
+    )
+
+    epc_mcs_df = pd.concat([epc_mcs_df, mcs_non_matched_df], axis=0)
+
+    if verbose:
+        print("EPC and MCS merged", epc_mcs_df.shape)
+
+    # Update installation date (will only affect MCS-only records)
+    epc_mcs_df["HP_INSTALL_DATE"] = np.where(
+        epc_mcs_df["MCS_AVAILABLE"],
+        epc_mcs_df["commission_date"],
+        epc_mcs_df["HP_INSTALL_DATE"],
+    )
+
+    # Update installation tag (will only affect MCS-only records)
+    epc_mcs_df["HP_INSTALLED"] = np.where(
+        epc_mcs_df["MCS_AVAILABLE"], True, epc_mcs_df["HP_INSTALLED"]
+    )
+
+    epc_mcs_df = standardise_hp_type(epc_mcs_df)
+
+    return epc_mcs_df
+
+
+def standardise_hp_type(epc_mcs_df):
+    """Standardise HP type between EPC and MCS data.
+
+    Args:
+        epc_mcs_df (pd.DataFrame): Merged EPC and MCS dataframe.
+
+    Returns:
+        pd.DataFrame: Merged EPC and MCS dataframe with standardised HP type.
+    """
+
+    type_dict = {
+        "No HP": "No HP",
+        "air source heat pump": "Air Source Heat Pump",
+        "ground source heat pump": "Ground/Water Source Heat Pump",
+        "heat pump": "Undefined or Other Heat Pump Type",
+        "water source heat pump": "Ground/Water Source Heat Pump",
+        "community heat pump": "Undefined or Other Heat Pump Type",
+        "Exhaust Air Heat Pump": "Undefined or Other Heat Pump Type",
+        "Air Source Heat Pump": "Air Source Heat Pump",
+        "Ground/Water Source Heat Pump": "Ground/Water Source Heat Pump",
+    }
+
+    # Overwrite with tech type from MCS if MCS data exists
+    epc_mcs_df["HP_TYPE"] = np.where(
+        epc_mcs_df["MCS_AVAILABLE"], epc_mcs_df["tech_type"], epc_mcs_df["HP_TYPE"]
+    )
+
+    epc_mcs_df["HP_TYPE"] = epc_mcs_df["HP_TYPE"].map(type_dict)
+
+    return epc_mcs_df
+
+
+def add_mcs_installer_data(epc_mcs_installations):
+    """Add MCS installer data to joined EPC and MCS installations dataframe based on installer ID.
+
+    Args:
+        epc_mcs_installations (pd.DataFrame): EPC and MCS installations data.
+    """
+
+    newest_hist_inst_batch = data_batches.get_latest_hist_installers()
+
+    print(newest_hist_inst_batch)
+
+    # # Load MCS
+    mcs_inst_data = data_getters.load_s3_data(
+        base_config.BUCKET_NAME,
+        newest_hist_inst_batch,
+    )
+
+    mcs_inst_data.rename(columns={"postcode": "POSTCODE"}, inplace=True)
+
+    # TODO Code to merge data with epc_mcs_installations based on company_unique_id
+    # It should be an outer merge keeping, keeping installer data with no matches to MCS installations
+    # Their MCS installations and EPC fields would be NaN
```

Comment:
Should be something like (but haven't tested running it):
```suggestion
    epc_mcs_installations = epc_mcs_installations.merge(
    right = newest_hist_inst_batch,
    how = "outer",
    left_on = ["company_unique_id", "installer_name"],
    right_on = ["company_unique_id", "company_name"])
    
   
   return epc_mcs_installations
```
Something still missing is the variables we want to get from the installers, to add to the merged dataset. Again, as I said above, I would probably pass it as an argument.

Reply:
Thanks a lot for this code snippet. Included it and will test.
I added usecols options for all datasets now.